### DISCUSSION TITLE
Prevent mixing of controlled & uncontrolled in custom components

### DISCUSSION THREAD

**Discussion by 7iomka** [mantinedev/mantine#2635] [Q&A]
Hi
There is a custom component QuantityInput
It uses recommendations from the docs, how to configure it to add support for controlled & uncontrolled version of usage.
I have one problem.
In all Mantine components, `value` & `onChange` props is a signal that component is controlled.
If one of this props is omitted (for example value is used without onChange) - value will be unchanged (see simple input in example).
However, in example below, if we used both value and onChange, but onChange not update value, internally logic to set value is mixed, and in UI we see that value changes.
How to prevent this unexpected behavior?
https://codesandbox.io/s/silent-bush-5l6l6g?file=/src/quantity-input/quantity-input.component.tsx

Thanks.

**Comment by rtivital** [comment]
I did not understand the issue or, most likely it is related only to NumberInput (because it supports undefined as value). Components are considered controlled when `value !== undefined`, onChange function is triggered without any conditions.

**Reply by 7iomka** [reply]
> I did not understand the issue or, most likely it is related only to NumberInput (because it supports undefined as value). Components are considered controlled when `value !== undefined`, onChange function is triggered without any conditions.

If NumberInput is an edge case, it is possible to prevent switching internally to uncontrolled behavior when value is passed?
In my example value is not undefined, but I want to prevent changing value internally.

**Reply by rtivital** [reply]
In your sandbox it does not work because `value` is an object (see console logs), I'm not sure where it is comming from – https://codesandbox.io/s/vigilant-sutherland-pene27?file=/src/quantity-input/quantity-input.component.tsx

**Reply by 7iomka** [reply]
> In your sandbox it does not work because `value` is an object (see console logs), I'm not sure where it is comming from – https://codesandbox.io/s/vigilant-sutherland-pene27?file=/src/quantity-input/quantity-input.component.tsx

Value is correct, it's not an object
You see wrong
<img width="1795" alt="Снимок экрана 2022-10-05 в 20 12 43" src="https://user-images.githubusercontent.com/18473137/194121272-9e411586-5dff-45d6-a6f4-c3dd948b2f86.png">

**Reply by rtivital** [reply]
I do not get what the issue is then, it works correctly. What did you expect from this sandbox?

**Reply by 7iomka** [reply]
> I do not get what the issue is then, it works correctly. What did you expect from this sandbox?

I expect to not see the value changes, because sync. with source (from useState) is broken, and handler not change source.
In other words, I want exact the same behavior like with Input component - you don't see any value changes in UI in the same case.

**Reply by rtivital** [reply]
Alright, now I get it. In this case, I would recommend managing state on your side to fix this issue for now.

**Reply by 7iomka** [reply]
> Alright, now I get it. In this case, I would recommend managing state on your side to fix this issue for now.

In other words, I can't make it hybrid - only controlled version is possible?

**Reply by rtivital** [reply]
This must be the case for NumberInput
